### ISSUE TITLE
UnicodeDecodeError: 'utf-8'

### ISSUE THREAD

**Issue by fariza** [mbr/latex#4]
Running the example

````
min_latex = (r"\documentclass{article}"
             r"\begin{document}"
             r"Hello, world!"
             r"\end{document}")

from latex import build_pdf

# this builds a pdf-file inside a temporary directory
pdf = build_pdf(min_latex)

# look at the first few bytes of the header
print (str(pdf)[:10])
````

I get 
```
Traceback (most recent call last):
  File "/home/fariza/workspace/Scripts/temp/pdf.py", line 27, in <module>
    print (str(pdf)[:10])
  File "/usr/local/lib/python3.4/dist-packages/data/__init__.py", line 125, in __str__
    return self.__unicode__()
  File "/usr/local/lib/python3.4/dist-packages/data/__init__.py", line 133, in __unicode__
    return self.file.read()
  File "/usr/lib/python3.4/codecs.py", line 313, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd0 in position 10: invalid continuation byte
```

As you can see, I am running in a Python3.4, any suggestions to make this work?

**Comment by mbr** [comment]
Sorry, this is an oversight due to Python2/3 incompatibilities.

In Python2, raw bytes are represented by the ``str``-type and unicode strings by the ``unicode`` type. In Python3, the ``unicode``-type has been renamed to ``str`` and for raw bytes, ``bytes`` has been introduced.

Since Python2.6, ``bytes`` is also available in Python2, but is just an alias for ``str``:

```
Python 2.6.9 (unknown, Apr 16 2015, 02:25:58) 
[GCC 4.2.1 Compatible Debian Clang 3.2 (tags/RELEASE_32/final)] on linux3
Type "help", "copyright", "credits" or "license" for more information.
>>> bytes
<type 'str'>
>>> 
```

That means ``bytes`` will give the correct type in both versions. To fix, replace

```python
print (str(pdf)[:10])
```

with

```python
print(bytes(pdf)[:10])
```

I've added a test for the example and fixed it in a2b6738. Hope this helps.

**Comment by fariza** [comment]
I just tried and get the same error

Also if I try to save the pdf
```
min_latex = (r"\documentclass{article}"
             r"\begin{document}"
             r"Hello, world!"
             r"\end{document}")
 
from latex import build_pdf
# 
# # this builds a pdf-file inside a temporary directory
pdf = build_pdf(min_latex)
# 
# # look at the first few bytes of the header
# print (bytes(pdf)[:10])
pdf.save_to('sample.pdf')
````

```
raceback (most recent call last):
  File "/home/fariza/workspace/Scripts/temp/pdf.py", line 47, in <module>
    pdf.save_to('sample.pdf')
  File "/usr/local/lib/python3.4/dist-packages/data/__init__.py", line 234, in save_to
    return self.save_to(out)
  File "/usr/local/lib/python3.4/dist-packages/data/__init__.py", line 226, in save_to
    dest.write(self.__bytes__())
  File "/usr/local/lib/python3.4/dist-packages/data/__init__.py", line 93, in __bytes__
    return self.file.read().encode(self.file.encoding)
  File "/usr/lib/python3.4/codecs.py", line 313, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd0 in position 10: invalid continuation byte
```

**Comment by mbr** [comment]
At this point, I have issues reproducing it:

```sh
$ mkvirtualenv broken-latex-py3 --python=python3.4
Running virtualenv with interpreter /home/marc/bin/python3.4
Using base prefix '/home/marc/Documents/github.com/pyenv/versions/3.4.3'
New python executable in broken-latex-py3/bin/python3.4
Also creating executable in broken-latex-py3/bin/python
Installing setuptools, pip...done.
$ workon broken-latex-py3
$ cd /tmp
$ mkdir TEST 
$ cd TEST
$ pip install latex
[...]
Collecting data (from latex)
  Using cached data-0.3.tar.gz
[...]
Collecting latex
  Downloading latex-0.5.3.tar.gz
$ python --version
$  cat > EX2.py
min_latex = (r"\documentclass{article}"
             r"\begin{document}"
             r"Hello, world!"
             r"\end{document}")

from latex import build_pdf
# 
# # this builds a pdf-file inside a temporary directory
pdf = build_pdf(min_latex)
# 
# # look at the first few bytes of the header
print (bytes(pdf)[:10])
$ python EX2.py
b'%PDF-1.5\n%'
$ cat > EX3.py
min_latex = (r"\documentclass{article}"
             r"\begin{document}"
             r"Hello, world!"
             r"\end{document}")

from latex import build_pdf
# 
# # this builds a pdf-file inside a temporary directory
pdf = build_pdf(min_latex)
# 
# # look at the first few bytes of the header
# print (bytes(pdf)[:10])
pdf.save_to('sample.pdf')
$ python EX3.py
$ file sample.pdf
sample.pdf: PDF document, version 1.5
```

Can you please check that you do

* not have any stale ``.pyc`` files
* have the latest version of ``data`` and ``latex`` (I just released an updated ``latex`` package, but it has only documentation fixes).
* create a fresh virtualenv to try this in?

Also, the output of pip freeze -l would be useful

**Comment by fariza** [comment]
```
$ virtualenv-3.4 broken-latex-py3
Using base prefix '/usr'
New python executable in broken-latex-py3/bin/python3
Also creating executable in broken-latex-py3/bin/python
Installing setuptools, pip, wheel...done.
$ source broken-latex-py3/bin/activate
$ cd /tmp/
$ mkdir TEST
$ cd TEST/
$ pip install latex
Collecting latex
Collecting tempdir (from latex)
Collecting data (from latex)
Collecting shutilwhich (from latex)
Collecting future (from latex)
Collecting funcsigs (from data->latex)
  Using cached funcsigs-0.4-py2.py3-none-any.whl
Collecting six (from data->latex)
  Using cached six-1.9.0-py2.py3-none-any.whl
Collecting decorator (from data->latex)
Installing collected packages: tempdir, funcsigs, six, decorator, data, shutilwhich, future, latex
Successfully installed data-0.3 decorator-3.4.2 funcsigs-0.4 future-0.14.3 latex-0.5.3 shutilwhich-1.1.0 six-1.9.0 tempdir-0.6
$ python --version
Python 3.4.0
$  cat > EX2.py
min_latex = (r"\documentclass{article}"
             r"\begin{document}"
             r"Hello, world!"
             r"\end{document}")

from latex import build_pdf
# 
# # this builds a pdf-file inside a temporary directory
pdf = build_pdf(min_latex)
# 
# # look at the first few bytes of the header
print (bytes(pdf)[:10])

$ python EX2.py
Traceback (most recent call last):
  File "EX2.py", line 12, in <module>
    print (bytes(pdf)[:10])
  File "/home/fariza/workspace/Scripts/temp/broken-latex-py3/lib/python3.4/site-packages/data/__init__.py", line 93, in __bytes__
    return self.file.read().encode(self.file.encoding)
  File "/home/fariza/workspace/Scripts/temp/broken-latex-py3/lib/python3.4/codecs.py", line 313, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd0 in position 10: invalid continuation byte
```

```
$ pip freeze -l 
data==0.3
decorator==3.4.2
funcsigs==0.4
future==0.14.3
latex==0.5.3
shutilwhich==1.1.0
six==1.9.0
tempdir==0.6
wheel==0.24.0

```

**Comment by mbr** [comment]
I'll check into this later this evening. It seems a bit strange.

**Comment by fariza** [comment]
Is there a way to keep all the intermediate files from the build process?

**Comment by mbr** [comment]
> Is there a way to keep all the intermediate files from the build process?

Not at this time. The main motivation was that I did *not* want them lying aronud =). I'll see if I can add that feature.

**Comment by mbr** [comment]
Finally fixed in a7d0136fe8e47c6c57dfd1ba3fd43614c2431f9e.
